### ISSUE TITLE
filters.litree routine stuck on first tree

### ISSUE THREAD

**Issue by jtenny** [PDAL/PDAL#3518]
**Description**
I made a pipeline to 
1. read from an EPT source
2. get height above ground with `filters.hag_delauney`
3. apply the Li tree segmentation routine (`filters.litree`)
4. write a .las file
 
I'm trying to use this on small areas, but everywhere I try, the Li segmentation routine always gets stuck on the first tree. Can't figure out what is wrong here, looks like a bug? Saw another user on stack exchange who might have the [same issue](https://gis.stackexchange.com/questions/396400/using-litree-in-pdal#comment658362_403125).

<details>
**Pipeline**

```
[
  {
    "type": "readers.ept",
    "filename": "https://s3-us-west-2.amazonaws.com/usgs-lidar-public/USGS_LPC_CA_NoCAL_Wildfires_B1_2018/ept.json",
    "bounds": "([-13444124.366465904, -13443980.986961761], [4772585.959571438, 4772718.552748669])"
  },
  {
    "type": "filters.hag_delaunay"
  },
  {
    "type": "filters.litree",
    "min_points": 10,
    "min_height": 3,
    "radius": 100
  },
  {
    "type": "writers.las",
    "filename": "pdal_request_seg.las",
    "extra_dims": "all"
  }
]
```

**Command**
```
$ pdal pipeline C:/Users/Johnathan/Downloads/pdal_request_seg.json --debug 
```

**Result**
```
(pdal pipeline writers.las Info) writers.las: Writing dimension ClassFlags(uint8_t)  to LAS extra bytes.
(pdal pipeline writers.las Info) writers.las: Writing dimension HeightAboveGround(double)  to LAS extra bytes.
(pdal pipeline writers.las Info) writers.las: Writing dimension TreeID(int64_t)  to LAS extra bytes.
(pdal pipeline Debug) Executing pipeline in standard mode.
(pdal pipeline filters.litree Debug) Classifying points for tree 1 (|Ui| = 107543)
```

**System Info**

PDAL Version: 2.2.0 (git-version: Release), installed through Conda
System: Microsoft Windows [Version 10.0.18363.836]

**Comment by jtenny** [comment]
@chambbj Looks like you implemented the `filters.litree` routine? Would you be able to try this pipeline and see if it works for you or if you are able to debug it?

**Comment by chambbj** [comment]
If I remember correctly, it is a pretty inefficient technique, at least my
implementation of it. I can look to see if there are any obvious issues,
but I can’t commit to full debugging at this time. If someone wants to fix
and provide a patch we would of course consider merging it, or if someone
wants to do the profiling/debugging that helps to identify and isolate the
issue, I’d be more inclined to look more closely. It would be great if we
can generate some working example test cases as these are clearly missing
currently.

**Comment by abellgithub** [comment]
Are you sure this isn't just running slowly?  I would try this:

- Remove the delaunay filter, as I don't think it's necessary for litree.
- Change the radius to something much smaller as it is currently including most of the points, if I don't misread.

If you're still having a problem, run in the debugger and halt the program after a bit and get a copy of the stack.

**Comment by jtenny** [comment]
Thank you both for the speedy responses.

- I don't think it's just slow. The area in this example is less than 2 hectares and this takes maybe 1 minute using the Li algorithm implemented in LidR (R studio). I let PDAL run overnight, about 12 hours and was still stuck on the debug line that says something like `(pdal pipeline filters.litree Debug) Classifying points for tree 1 (|Ui| = 107543)` when I got back.
- The filter does seem to require height above ground, not sure if there is a better way to get it than delauney
- Experimented with radii 25 and 2, leaving it for 15+ minutes each time, still no dice.

I'd love to use this filter in PDAL, it would definitely simplify my workflow, but it sounds like I'll need to stick with LidR for now. I also don't have much time for debugging right now and my coding skills are pretty weak anyways. If someone is able to put some more work into this and get an example working with this dataset, I'd be happy to jump back in for some additional testing.

**Comment by chambbj** [comment]
@jtenny Thanks for bringing up this issue and providing a clear, reproducible pipeline.

After speaking with @abellgithub and taking a closer look, I was able to identify a number of opportunities for improvement.

Here's an example of the new output, which I was able to produce in a little over a minute (we can probably get that time down even further, I suspect).

![newliresult](https://user-images.githubusercontent.com/717558/129416024-bc5fbd3d-e82f-4172-9710-c4b9daed35a0.png)

I'll need to clean this up a bit first, but will then submit a pull request resolving the multiple issues and we'll ship it with the next release.
